packages/block-editor/src/store/reducer.js

```diff
@@ -239,6 +224,37 @@ function withPersistentBlockChange( reducer ) {
 	};
 }
 
+/**
+ * Higher-order reducer intended to augment the blocks reducer, assigning an
+ * `isIgnoredChange` property value corresponding to whether a change in state
+ * can be considered as ignored. A change is considered ignored when the result
+ * of an action not incurred by direct user interaction.
+ *
+ * @param {Function} reducer Original reducer function.
+ *
+ * @return {Function} Enhanced reducer function.
+ */
+function withIgnoredBlockChange( reducer ) {
+	/**
+	 * Set of action types for which a blocks state change should be ignored.
+	 *
+	 * @type {Set}
+	 */
+	const IGNORED_ACTION_TYPES = new Set( [
+		'RECEIVE_BLOCKS',
+	] );
+
+	return ( state, action ) => {
+		const nextState = reducer( state, action );
+
+		if ( nextState !== state ) {
+			nextState.isIgnoredChange = IGNORED_ACTION_TYPES.has( action.type );
```

Comment:
so what is exactly the difference between isIgnoredChange and ! isPersistentChange?

Reply:
From `__unstableIsLastBlockChangeIgnored` docs:

>An ignored change is one not to be committed by BlockEditorProvider, neither via `onChange` nor `onInput`.

From `isLastBlockChangePersistent` docs:

>A persistent change is one committed by BlockEditorProvider via its `onChange` callback, in addition to `onInput`.

Effectively: Non-persistent changes are still surfaced up to the rendering Editor, but treated as omitted from history. Ignored changes are not surfaced to the rendering Editor at all.

Reply:
Oh I think I got it, while one is about the `onChange` vs `onInput` the other is about not calling them entirely.
It does seems like they have the same roles though. Avoid undo levels, can't we unify and just use one?

Reply:
(I replied to myself before I see your comment)

Reply:
>It does seems like they have the same roles though. Avoid undo levels, can't we unify and just use one?

This is a good summary of the problem. They're not the same. One is considered by `withHistory`, the other by `withChangeDetection`. The bug exists because we only accounted for the first, not the second.

If the question is whether "changes" should be a reflection of the presence of undo levels: I think long ago we had it this way, and it changed (presumably necessarily). Possibly as a result of some of the work to merge undo levels while still needing to consider them as changed.

Reply:
👍 Thanks for the clarification, it makes sense.

Reply:
> If the question is whether "changes" should be a reflection of the presence of undo levels: I think long ago we had it this way, and it changed (presumably necessarily). Possibly as a result of some of the work to merge undo levels while still needing to consider them as changed.

I think the reason we don't consider Undo levels as part of post dirtiness is because Undo history doesn't reset when a Save occurs, but the post is "clean".